subprojects/gst-plugins-base/gst-libs/gst/gl/web/gstglwindow_canvas.c

```diff
@@ -0,0 +1,172 @@
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <emscripten/html5.h>
+#include <gst/gst.h>
+#include <gst/gl/gl.h>
+#include "gstglcontext_emscripten.h"
+#include "gstglwindow_canvas.h"
+#include "../gstglwindow_private.h"
+
+#define GST_CAT_DEFAULT gst_gl_window_debug
+
+#define gst_gl_window_canvas_parent_class parent_class
+G_DEFINE_TYPE (GstGLWindowCanvas, gst_gl_window_canvas,
+    GST_TYPE_GL_WINDOW);
+
+static EM_BOOL
+gst_gl_window_canvas_mouse_cb (int event_type,
+    const EmscriptenMouseEvent *event, void *data)
+{
+  GstGLWindow *window = GST_GL_WINDOW (data);
+  GstGLWindowCanvas *self = GST_GL_WINDOW_CANVAS (window);
+
+  GST_DEBUG_OBJECT (self, "Mouse event %d received", event_type);
+  switch (event_type) {
+    case EMSCRIPTEN_EVENT_MOUSEMOVE:
+      GST_LOG_OBJECT (window, "Mouse move to %d %d", event->targetX,
+          event->targetY);
+      gst_gl_window_send_mouse_event (window, "mouse-move", 0, event->targetX,
+          event->targetY);
+      break;
+    case EMSCRIPTEN_EVENT_MOUSEDOWN:
+      GST_LOG_OBJECT (window, "Mouse down with button %d at %d %d",
+          event->button, event->targetX, event->targetY);
+      gst_gl_window_send_mouse_event (window, "mouse-button-press",
+          event->button, event->targetX, event->targetY);
+      break;
+    case EMSCRIPTEN_EVENT_MOUSEUP:
+      GST_LOG_OBJECT (window, "Mouse up with button %d at %d %d",
+          event->button, event->targetX, event->targetY);
+      gst_gl_window_send_mouse_event (window, "mouse-button-release",
+          event->button, event->targetX, event->targetY);
+      break;
+    default:
+      GST_WARNING_OBJECT (self, "Event %d not handled", event_type);
+      break;
+  }
+  return EM_TRUE;
+}
+
+static void
+gst_gl_window_canvas_init (GstGLWindowCanvas * self)
+{
+}
+
+static guintptr
+gst_gl_window_canvas_get_window_handle (GstGLWindow * window)
+{
+  GstGLWindowCanvas *self = GST_GL_WINDOW_CANVAS (window);
+
+  return (guintptr) self->canvas;
+}
+
+static void
+gst_gl_window_canvas_draw_cb (gpointer data)
+{
+  GstGLWindowCanvas *self = data;
+  GstGLWindow *window = GST_GL_WINDOW (self);
+  GstGLContext *context = gst_gl_window_get_context (window);
+  guintptr context_handle;
+  gint width, height;
+  guint window_width, window_height;
+
+  /* Given that we can not register for a callback of when the canvas
+   * sized changed, we need to ask for it ourselves
+   */
+
+  gst_gl_window_get_surface_dimensions (window, &window_width,
+      &window_height);
+  context_handle = gst_gl_context_get_gl_context (context);
+  if (!emscripten_webgl_get_drawing_buffer_size (context_handle,
+          &width, &height)
+      && (window->queue_resize || width != window_width
+          || height != window_height)) {
+    GST_DEBUG_OBJECT (window, "Resizing to %dx%d from %dx%d", width, height,
+        window_width, window_height);
+    gst_gl_window_resize (window, width, height);
+  }
+
+  if (window->draw)
```

Comment:
`window->draw` will be always defined, right?

Reply:
Not really. It is an instance vmethod defined for every GstGLWindow the user can hook to. Not having it is not a problem.